### PR TITLE
DOCK-2600: Fix backlinks from Zenodo DOI records

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
As documented in the ticket, we recently discovered that sometimes, when we create a Zenodo DOI record, some of the backlinks from the Zenodo record to the entry do not work.  Specifically, the TRS URL and alias URL backlinks are invalid for non-BioWorkflow entries, because the both should contain a substring that denotes the type of entry (`notebook`, `service`, etc), and instead, in the broken URLs in the Zenodo record, the substring `workflow` appears for all entry types.  This problem probably started when we did not adjust the responsible code to handle services, and continued, unmarked by service mods (which I used to pinpoint where to modify the codebase for notebooks, for example), as we added more entry types.

This PR fixes the two responsible code sites, moves the alias URL generation into a helper, and tests both types of URL generation for both BioWorkflows and Notebooks.  Also, we streamline `ZenodoHelper.createWorkflowTrsUrl` by changing it to `Entry.getTrsId` and shortening and clarifying some variable names.  Coulda kept on refactoring but stopped myself... :) `ToolsImplCommon.getUrl` (and related) probably belong in a `URLHelper`-type class.

As an aside, if we're pondering maybe eventually supporting freeform entries that may not have a single, well-defined type, the entry type in these URLs could theoretically be problematic.  Would it be better to reference versions (and entries?) without involving the entry type?  Likely pretty far off in the Dockstore evolution, but food for thought...

**Review Instructions**
Check the Zenodo page pointed at by a Dockstore-generated DOI, and confirm that all of the backlinks resolve to the appropriate content.

**Issue**
https://ucsc-cgl.atlassian.net/browse/DOCK-2600
#6033

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 

- [x] Security and Privacy assessed

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by svonworl** [comment]
Woops, didn't start this off the hotfix branch, will probably be easiest to merge this to develop and then cherry pick to hotfix branch in another PR.

**Comment by denis-yuen** [comment]
> As an aside, if we're pondering maybe eventually supporting freeform entries that may not have a single, well-defined type, the entry type in these URLs could theoretically be problematic. Would it be better to reference versions (and entries?) without involving the entry type? Likely pretty far off in the Dockstore evolution, but food for thought...

If I understand correctly, for the TRS IDs, I think we started with tools not having an entry type and then workflows got `#workflow` to distinguish them.

**Review by denis-yuen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AliasHelper.java [hunk 1]
```diff
@@ -96,4 +96,8 @@ public static WorkflowVersion addWorkflowVersionAliasesAndCheck(AuthenticatedRes
         newAliases.forEach(alias -> workflowVersion.getAliases().put(alias, new Alias()));
         return workflowVersion;
     }
+
+    public static String createWorkflowVersionAliasUrl(String dockstoreUrl, Workflow workflow, String alias) {
+        return "%s/aliases/%s-versions/%s".formatted(dockstoreUrl, workflow.getEntryTypeMetadata().getTerm(), alias);
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AliasHelper.java:101]: In the web service, t looks like there is only a path for `...workflow-versions...`, [here](https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasResource.java#L74).

Will the UI convert `notebook-versions` to `workflow-versions` when it invokes the Dockstore API?

Trying to make sure I understand correctly.

**Review by coverbeck — COMMENTED** [review]
Asked a question for now to make sure I understand.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AliasHelper.java (hunk 1 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AliasHelper.java:101]: Yes, the URL created by `createWorkflowVersionAliasUrl` references our site, where UI code retrieves information about the alias via the API call you mention, and then redirects.  The latter magic happens in `aliases/aliases.component.ts` and related.

**Review by coverbeck — APPROVED** [review]
> Sort of. The URL created by createWorkflowVersionAliasUrl references our site, where UI code retrieves information about the alias via the API call you mention, and then redirects. The latter magic happens in aliases/aliases.component.ts and related.

It still feels a little weird to me that the notebook version DOI has a different path segment than fetching the version via the the DOI with the API. But as the links will now work, I think it's OK.

**Review by kathy-t — APPROVED** [review]

**Comment by svonworl** [comment]
> > Sort of. The URL created by createWorkflowVersionAliasUrl references our site, where UI code retrieves information about the alias via the API call you mention, and then redirects. The latter magic happens in aliases/aliases.component.ts and related.
> 
> It still feels a little weird to me that the notebook version DOI has a different path segment than fetching the version via the the DOI with the API. But as the links will now work, I think it's OK.

Certainly, the two are not consistent, but to make them so (by only changing one or the other), we'd either need to change the way aliases work (make aliases containing `/workflow-versions/` resolve to all non-old-style-tool entry types, which would be a strange way to link to a notebook, for example), or add a new endpoints that are the analog to the alias `/workflow-versions/` endpoint for each of the entry types.  Neither is particularly palatable/practical.

**Comment by svonworl** [comment]
> > > Sort of. The URL created by createWorkflowVersionAliasUrl references our site, where UI code retrieves information about the alias via the API call you mention, and then redirects. The latter magic happens in aliases/aliases.component.ts and related.
> > 
> > 
> > It still feels a little weird to me that the notebook version DOI has a different path segment than fetching the version via the the DOI with the API. But as the links will now work, I think it's OK.
> 
> Certainly, the two are not consistent, but to make them so (by only changing one or the other), we'd either need to change the way aliases work (make aliases containing `/workflow-versions/` resolve to all non-old-style-tool entry types, which would be a strange way to link to a notebook, for example), or add a new endpoints that are the analog to the alias `/workflow-versions/` endpoint for each of the entry types. Neither is particularly palatable/practical.

This touches on the observation that maybe it's not always a good thing, from a design point-of-view, for the entry type to be part of the url/path.  If they weren't in there, the inconsistencies wouldn't exist...

**Pull request merged**